crates/uv-auth/src/middleware.rs

```diff
@@ -0,0 +1,185 @@
+use netrc::Netrc;
+use reqwest::{header::HeaderValue, Request, Response};
+use reqwest_middleware::{Middleware, Next};
+use std::path::Path;
+use task_local_extensions::Extensions;
+use tracing::{debug, warn};
+
+use crate::{
+    keyring::{get_keyring_subprocess_auth, KeyringProvider},
+    store::{AuthenticationStore, Credential},
+};
+
+/// A middleware that adds basic authentication to requests based on the netrc file and the keyring.
+///
+/// Netrc support Based on: <https://github.com/gribouille/netrc>.
+pub struct AuthMiddleware {
+    nrc: Option<Netrc>,
+    keyring_provider: KeyringProvider,
+}
+
+impl AuthMiddleware {
+    pub fn new(keyring_provider: KeyringProvider) -> Self {
+        Self {
+            nrc: Netrc::new().ok(),
+            keyring_provider,
+        }
+    }
+
+    pub fn from_netrc_file(file: &Path, keyring_provider: KeyringProvider) -> Self {
+        Self {
+            nrc: Netrc::from_file(file).ok(),
+            keyring_provider,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Middleware for AuthMiddleware {
+    async fn handle(
+        &self,
+        mut req: Request,
+        _extensions: &mut Extensions,
+        next: Next<'_>,
+    ) -> reqwest_middleware::Result<Response> {
+        let url = req.url().clone();
+        // If the request already has an authorization header, we don't need to do anything.
+        // This gives in-URL credentials precedence over the netrc file.
+        if req.headers().contains_key(reqwest::header::AUTHORIZATION) {
+            if !url.username().is_empty() {
+                AuthenticationStore::save_from_url(&url);
+            }
+            return next.run(req, _extensions).await;
+        }
+
+        // Try auth strategies in order of precedence:
+        if let Some(stored_auth) = AuthenticationStore::get(&url) {
+            // If we've already seen this URL, we can use the stored credentials
+            if let Some(auth) = stored_auth {
+                match auth {
+                    Credential::Basic(_) => {
+                        req.headers_mut().insert(
+                            reqwest::header::AUTHORIZATION,
+                            basic_auth(auth.username(), auth.password()),
+                        );
+                    }
+                    // Url must already have auth if before middleware runs - see `AuthenticationStore::with_url_encoded_auth`
+                    Credential::UrlEncoded(_) => (),
```

Comment:
Can you expand on the comment here? Should this be `unreachable!`?

Reply:
The idea here is, if the url does not come to the middleware with URL-encoded auth, we shouldn't introduce URL-encoded auth because it's an inherently unsafe way to send auth data.  On main, there was already code copying the auth from trusted urls onto response/body urls so this line shouldn't be hit under normal cicumstances anyways.